Add support for linear in mkldnn fusion

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* **#51484 Add support for linear in mkldnn fusion**
* #51483 Move conv to mkldnn

This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .

Differential Revision: [D26696698](https://our.internmc.facebook.com/intern/diff/D26696698)

Commit:
Add support for linear

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -138,66 +162,89 @@ bool frozenMkldnnCompatibleConvNode(Node* n) {
+  replaceInputWithMKLDNNTensor(n, name, mkldnn_tensor);
+}
+
+void convertFrozenConvToMKLDNN(Node* conv) {
```
did anything change in here from the previous PR?

eellison replied:
No, just refactoring..

Krovatkin commented on test/jit/test_freezing.py:
```diff
@@ -1563,6 +1563,48 @@ def to_dense(inp):
+                super().__init__(*args, **kwargs)
+
+            def forward(self, input):
+                return torch._C._nn.linear(input, self.weight, self.bias)
```
this is to get rid of an if?

eellison replied:
This is to compile it as one op. This is how it's done on master as of now but wasnt landed yet when i submitted this PR.

Krovatkin commented on aten/src/ATen/native/mkldnn/Linear.cpp:
```diff
@@ -28,16 +28,22 @@ Tensor mkldnn_linear(
-      "mkldnn_linear: input needs to has dim at least 2, input dim ", self.dim());
+  const int64_t dim = self.dim();
+  TORCH_CHECK(
+      self.dim() != 0,
```
I wonder if there was a performance concern here. 
MKLDNN also has `inner_product`: https://oneapi-src.github.io/oneDNN/structmkldnn_1_1inner__product__forward_1_1desc.html#details We could ask Intel guys.

eellison replied:
Yea, that's a good idea. With dim > 2 they just do a reshape, so this is equivalent to that approach. From looking at the top 200 shapes internally broadcasting linear is extremely rare. I think `inner_product` might be a good follow up but not necessary for this PR..

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "[WIP] Add support for linear"




[ghstack-poisoned]

Commit:
Update on "[WIP] Add support for linear"




[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .







[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .







[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .







[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .







[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .







[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .







[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .







[ghstack-poisoned]

Commit:
Update on "Add support for linear in mkldnn fusion"


This PR moves the linear weights of a frozen model to MKLDNN. When the weights are already in MKLDNN, just computing a single linear by converting the input and output from/to mkldnn provides large speedups. I benchmark'd the results of the top 200 shapes in predictor [here](https://www.internalfb.com/phabricator/paste/view/P171537854) (taken from aten::matmul), as well as verified that it sped up popular models. .

Differential Revision: [D26696698](https://our.internmc.facebook.com/intern/diff/D26696698)

[ghstack-poisoned]
